File: ml-benchmark/synduce/constraints/unimodal_lists/max_logn_v2.f
Inductive UList = uelt Int | usplit {UList, Int, Int, UList};
Inductive List = elt Int | cons {Int, List};

repr =
let aux = (fix (
  \f: List -> UList -> List. \res: List. \xs: UList.
  match xs with
    uelt x -> cons {x, res}
  | usplit {x, a, b, y} -> f (cons {a, cons {b, f res y}}) x
  end
)) in fix (
  \f: UList -> List. \xs: UList.
  match xs with
    uelt x -> elt x
  | usplit {x, a, b, y} -> aux (cons {a, cons {b, f y}}) x
  end
);

is_unimodal =
  let aux_down = fix (
    \f: Int -> List -> Bool. \pre: Int. \xs: List.
    match xs with
      elt x -> > pre x
    | cons {h, t} -> and (> pre h) (f h t)
    end
  ) in let aux_up = fix (
    \f: Int -> List -> Bool. \pre: Int. \xs: List.
    match xs with
      elt x -> < pre x
    | cons {h, t} -> if < pre h then f h t else aux_down h t
    end
  ) in \xs: List.
  match xs with
    elt x -> true
  | cons {h, t} -> aux_up h t
  end;

max = \a: Int. \b: Int. if < a b then b else a;

spec = fix (
  \f: List -> Int. \xs: List.
  match xs with
    elt x -> x
  | cons {h, t} -> max h (f t)
  end
);

target = fix (
  \f: UList -> Compress UList. \xs: UList.
  match xs with
    uelt x -> uelt x
  | usplit {x, a, b, y} ->
    if > a b then usplit {f x, a, b, y} /*Avoid the recursion of y*/
    else usplit {x, a, b, f y} /*Avoid the recusion of x*/
  end
);

main = \xs: UList. if is_unimodal (repr xs) then spec (repr (target xs)) else 0;

File: ml-benchmark/synduce/constraints/unimodal_lists/max_logn_v2.ml
type ulist = Uelt of int | Usplit of ulist * int * int * ulist
type llist = Elt of int | Cons of int * llist

let rec aux res xs =
  match xs with
  | Uelt x -> Cons (x, res)
  | Usplit (x, a, b, y) ->
      let t0 = aux res y in
      let t1 = Cons (b, t0) in
      let t2 = Cons (a, t1) in
      aux t2 x

let rec repr xs =
  match xs with
  | Uelt x -> Elt x
  | Usplit (x, a, b, y) ->
      let t0 = repr y in
      let t1 = Cons (b, t0) in
      let t2 = Cons (a, t1) in
      aux t2 x

let rec aux_down pre xs =
  match xs with
  | Elt x -> x > pre
  | Cons (h, t) ->
      if h > pre then aux_down h t else false

let rec aux_up pre xs =
  match xs with
  | Elt x -> x < pre
  | Cons (h, t) ->
      if pre < h then aux_up h t else aux_down h t

let is_unimodal xs =
  match xs with
  | Elt _ -> true
  | Cons (h, t) -> aux_up h t

let max a b = if a < b then b else a

let rec spec xs =
  match xs with
  | Elt x -> x
  | Cons (h, t) ->
      let m = spec t in
      max h m

val target: ulist -> ulist compress
let rec target xs =
  match xs with
  | Uelt x -> Uelt x
  | Usplit (x, a, b, y) ->
      if a > b then
        let tx = target x in
        Usplit (tx, a, b, y)
      else
        let ty = target y in
        Usplit (x, a, b, ty)

let program xs =
  if is_unimodal (repr xs) then
    let t = target xs in
    let r = repr t in
    spec r
  else
    0